### PR TITLE
SSR Fixes and better tests for v3.0.0

### PR DESCRIPTION
**Pull request by Judahmeek**
### Summary

_Remove this paragraph and provide a general description of the code changes in your pull
request... were there any bugs you had fixed? If so, mention them. If
these bugs have open GitHub issues, be sure to tag them here as well,
to keep the conversation linked together._

### Other Information

_Remove this paragraph and mention any other important and relevant information such as benchmarks._

### Pull Request checklist
_Remove this line after checking all the items here. If the item is not applicable to the PR, both check it out and wrap it by `~`._

- [ ] Add/update test to cover these changes
- [ ] Update documentation
- [ ] Update CHANGELOG file

### CODE REVIEW

File: rakelib/create_release.rake [hunk 1]
```diff
@@ -39,7 +39,7 @@ task :create_release, %i[gem_version dry_run] do |_t, args|
   puts 'Updating ujs:update'
   Rake::Task['ujs:update'].invoke
 
-  Release.commit_the_changes('Update pre-bundled react and React ujs') unless is_dry_run
```
Comment (ahangarha) [inline rakelib/create_release.rake:42]: Then the `commit_the_changes` method should be removed

File: lib/react/rails/version.rb [hunk 2]
```diff
@@ -2,6 +2,6 @@ module React
   module Rails
     # If you change this, make sure to update VERSIONS.md
     # and republish the UJS by updating package.json and `bundle exec rake ujs:publish`
-    VERSION = '2.7.0.rc.2'
+    VERSION = '2.6.2'
```
Comment (ahangarha) [inline lib/react/rails/version.rb:5]: Why still 2.6.2?

File: .github/workflows/ruby.yml [hunk 3]
```diff
@@ -6,23 +6,39 @@
...
+      - uses: ruby/setup-ruby@v1
+        with:
+          bundler: 2.4.9
+          ruby-version: ${{ matrix.ruby }}
+      - uses: actions/setup-node@v3
+      - run: bundle install
+      - run: yarn
+      - run: bundle exec rake react:update
+      - run: bundle exec rake ujs:update
+      - run: ./check_for_uncommitted_files.sh
```
Comment (ahangarha) [inline .github/workflows/ruby.yml:28]: Maybe we can follow the DRY principle and create a rake task for this check. We can use it here and also in the release task.

**Review by ahangarha — COMMENTED** [review]
Very extensive PR 🙂 

It would be great if we could break it into two PRs
- Fix CI and the issue with requiring `react-dom/client` -> 2.7.1
- SSR fix -> 2.8 or 3.0

File: lib/react/rails/version.rb (hunk 2 above)
Comment (Judahmeek) [reply lib/react/rails/version.rb]: Fixed with rebase

File: Appraisals [hunk 4]
```diff
@@ -3,29 +3,18 @@ appraise 'rails-5.2-sprockets_4' do
...
   gem 'rails', '~> 5.2.x'
   gem 'sprockets', '~> 3.5'
   gem 'turbolinks', '~> 5'
   gem 'mini_racer', :platforms => :mri
-  gem 'therubyrhino', :platforms => :jruby
 end
 
-appraise 'rails-5.2_no_sprockets_webpacker_3' do
+appraise 'rails-5.2_no_sprockets_shakapacker' do
   gem 'rails', '~> 5.2.x'
```
Comment (justin808) [inline Appraisals:16]: how much longer should we support rails 5.x?

Rails 5.2.1 Ended 11 months ago, (01 Jun 2022)

File: test/support/webpacker_helpers.rb [hunk 5]
```diff
@@ -1,14 +1,9 @@
...
-  begin
-    MAJOR, MINOR, PATCH, _ = Bundler.locked_gems.specs.find { |gem_spec| gem_spec.name == 'webpacker' }.version.segments
-  rescue
-    MAJOR, MINOR, PATCH, _ = [0,0,0]
-  end
 
   module_function
   def available?
-    defined?(Webpacker)
+    !!defined?(Webpacker)
```
Comment (justin808) [inline test/support/webpacker_helpers.rb:6]: Should we also add `Shakapacker` to support v7+?

File: test/support/webpacker_helpers.rb [hunk 6]
```diff
@@ -21,10 +16,8 @@ def compile
     return unless available?
     clear_webpacker_packs
     Dir.chdir("./test/#{DUMMY_LOCATION}") do
-      capture_io do
-        Rake::Task['webpacker:compile'].reenable
-        Rake::Task['webpacker:compile'].invoke
-      end
+      Rake::Task['webpacker:compile'].reenable
```
Comment (justin808) [inline test/support/webpacker_helpers.rb:19]: what is this change?

**Review by justin808 — COMMENTED** [review]
Looking good. 

Let's get PR 1278 merged and fix the conflicts.

**Review by justin808 — COMMENTED** [review]
Looking good. 

Let's get PR 1278 merged and fix the conflicts.

File: Appraisals (hunk 4 above)
Comment (ahangarha) [reply Appraisals:16]: I recommend making a yearly plan for our support for different versions of Ruby, Rails, Node, and React. Then update documentation and CI in all of our projects based on that.

It should be done in a (maximum) 30 min call with key senior developers in the team.

File: test/support/webpacker_helpers.rb (hunk 6 above)
Comment (Judahmeek) [reply test/support/webpacker_helpers.rb:19]: I removed `capture_io`

File: lib/react/server_rendering/exec_js_renderer.rb [hunk 7]
```diff
@@ -10,7 +10,9 @@ class ExecJSRenderer
 
       def initialize(options={})
         js_code = options[:code] || raise('Pass `code:` option to instantiate a JS context!')
-        @context = ExecJS.compile(GLOBAL_WRAPPER + js_code)
+        full_code = GLOBAL_WRAPPER + js_code
+        # File.write("./test/dummy/tmp/latest_js_context.js", full_code)
```
Comment (justin808) [inline lib/react/server_rendering/exec_js_renderer.rb:14]: @Judahmeek keep this comment?

File: lib/react/server_rendering/separate_server_bundle_container.rb [hunk 8]
```diff
@@ -0,0 +1,18 @@
+require 'open-uri'
+
+module React
+  module ServerRendering
+    # Get a compiled file from Webpacker's output path
+    class SeparateServerBundleContainer
+
+      def self.compatible?
+        !!defined?(Webpacker)
```
Comment (justin808) [inline lib/react/server_rendering/separate_server_bundle_container.rb:9]: Check for `Shakapacker`?

**Review by justin808 — APPROVED** [review]
Looks ready. One debug comment left accidentally?

Do we need to update the README? and CHANGELOG? and release v3? Any upgrade guide?

CC: @ahangarha @Judahmeek

File: lib/react/server_rendering/exec_js_renderer.rb (hunk 7 above)
Comment (Judahmeek) [reply lib/react/server_rendering/exec_js_renderer.rb:14]: I had to use it to debug two different SSR error with ExecJS so yes, I'd like to keep it.

File: lib/react/server_rendering/separate_server_bundle_container.rb (hunk 8 above)
Comment (Judahmeek) [reply lib/react/server_rendering/separate_server_bundle_container.rb:9]: I don't want to try supporting Shakapacker v6 & v7 at the same time.

**Pull request merged**